Add section with guidelines for Slot Aggregators

Mr0grog opened the pull request:
After discussion with @jmandel on FHIR chat, I’m proposing a new section in the specification specifically oriented towards *Slot Aggregators*. There are a number of unusual situations aggregators encounter, but that should not apply to first-party slot publishers, and this section describes features and patterns they can use. It’s a separate section to more clearly separate it from the rest of the specification and discourage *Slot Publishers* from using these features.

I’ve tried to gather different approaches in use by folks at USDS, USDR, and VaccineSpotter or that have been discussed in chat, but this section is probably imperfect and could use some feedback.

/cc @jmandel @nickrobison-usds @GUI

Mr0grog commented on specification.md:
````diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+
+#### Unknown Availability
+
+If there is no source of appointment data for a Schedule or a source system is unreachable, has errors, or cannot be reliably handled for some reason, a _Slot Aggregator_ should publish a single slot for the Schedule with a `status` value of `"free"` and a “Capacity” extension with a value of `0`. First-party _Slot Publishers_ SHOULD NOT publish a slot with these features; if a slot has no free c ...
+
+Example usage:
+
+```json
+{
+  "resourceType": "Slot",
+  "id": "789",
+  "schedule": {
+    "reference": "Schedule/456"
+  },
+  "status": "free",
+  "start": "2021-03-10T00:00:00Z",
+  "end": "2021-03-24T00:00:00Z",
+  "extension": [{
+    "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/slot-capacity",
+    "valueInteger": 0
+  }]
+}
+```
````
This is what we’re currently doing at USDR, but this solution feels really awkward to me. I wonder if this would be better handled by using the “has availability” extension on the Schedule with no `valueBoolean` or by changing it to use `valueCode` instead (with codes like `"yes" | "no" | "unknown"`).

jmandel replied:
Ah, yes -- I like this. Might say `some`, `none` and `unknown` rather than "yes"/"no".

Mr0grog commented on specification.md:
```diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+
+#### Unknown Capacity or Slot Times
+
+_Slot Aggregators_ should use the **optional "Has Availability" extension** to convey that a Schedule has non-zero future availability, without conveying details about when or how much. When used on a Schedule, that Schedule MAY have no associated Slots. _Slot Publishers_ SHALL NOT use this capacity in place of publishing granular Slots; it is defined to support Slot Aggregators (i.e. systems tha ...
```
I’ve added a sentence in the middle here about a schedule using this not needing slots. I wrote “MAY have no associated Slots,” but that leaves things a bit ambiguous (you could use it as a shortcut/summary even when slot-level detail is available here). Should we change this to “SHALL” or, conversely, explicitly suggest that this can be used to summarize data in slots?

See also the two comments above:
1. Would it make sense to allow this on slots, too? It would describe the CVS-style case well, and now that this is in a separate section from the rest of the spec, it might run *less* afoul of worries that publishers would abuse it. Not a huge deal either way, as the abuse is definitely a concern.
2. If we allowed a no-value case (indicating unknown availability) or switched it from `valueBoolean` to `valueCode`, that would solve the “unknown availability” case in a pretty elegant way.

jmandel commented on specification.md:
```diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+
+### Data “Freshness”
+
+Aggregators may request or receive information from publishers at different times, and understanding how recently data about a Location, Schedule, or Slot was retrieved can help end users gauge the accuracy of items in an aggregated dataset. _Slot Aggregators_ SHOULD use the `lastSourceSync` extension on the `meta` field of any resource to indicate the last time at which the data was known to be  ...
```
I think it's fine for us to use it here; we're not exactly bringing this guide through a normative process yet, and by the time we want to, the `lastSourceSync` will probably be ready.

jmandel commented on specification.md:
```diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+      "value": "CV1654321"
+    },
+    {
+      "system": "https://flynnspharmacy.org/",
```
Can you add ".example.org" to any example domains? e.g. "flynnspharamcy.example.org"?

Mr0grog replied:
Oh, yes, that makes sense. Will do.

jmandel commented on specification.md:
```diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+
+### Unknown Availability, Capacity, or Slot Times
+
+Because source systems may experience errors or may not conform to the SMART Scheduling Links specification, _Slot Aggregators_ need additional tools to describe unusual situations that are not relevant to first-part _Slot Publishers_. Specifically, _Slot Aggregators_ may describe locations with:
```
```suggestion
Because source systems may experience errors or may not conform to the SMART Scheduling Links specification, _Slot Aggregators_ need additional tools to describe unusual situations that are not relevant to first-part _Slot Publishers_. Specifically, _Slot Aggregators_ may describe Schedules with:
```

Is "schedules" more accurate, or am I misunderstanding?

Mr0grog replied:
Yep, that would be more accurate. I got tied up in my COVID-centric use case. 😅

jmandel commented on specification.md:
```diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+### Unknown Availability, Capacity, or Slot Times
+
+Because source systems may experience errors or may not conform to the SMART Scheduling Links specification, _Slot Aggregators_ need additional tools to describe unusual situations that are not relevant to first-part _Slot Publishers_. Specifically, _Slot Aggregators_ may describe locations with:
+- Unknown slot availability (e.g. because a source system is unreachable),
```
```suggestion
- Unknown whether any Slots are associated with this Schedule (e.g. because a source system is unreachable),
```

jmandel commented on specification.md:
```diff
@@ -344,3 +337,183 @@ In this case, if the `source` value is `source-abc` and the `booking-referral` i
+
+#### Unknown Availability
+
+If there is no source of appointment data for a Schedule or a source system is unreachable, has errors, or cannot be reliably handled for some reason, a _Slot Aggregator_ should publish a single slot for the Schedule with a `status` value of `"free"` and a “Capacity” extension with a value of `0`. First-party _Slot Publishers_ SHOULD NOT publish a slot with these features; if a slot has no free c ...
```
Why no mark "nothing is known about any Slots" as an extension on the *Schedule*? I thought that was what we had discussed, but I might be mis-remembering.

jmandel reviewed (changes requested):
This is great -- I think the simplification of using ternary logic on the "has-availability" extension is a good idea.

Mr0grog commented:
OK! I think I addressed all the issues you brought up, @jmandel. The one outstanding question here is:

> Would it make sense to allow [has-availability] on slots, too? It would describe the CVS-style case well [even if not for use by CVS], and now that this is in a separate section from the rest of the spec, it might run less afoul of worries that publishers would abuse it. Not a huge deal either way, as the abuse is definitely a concern.
